website/content/docs/pipelines/index.mdx

```diff
@@ -0,0 +1,300 @@
+security scan the built artifact.
+
+```hcl
+pipeline "example" {
+  step "run-some-code" {
+    image_url = "https://localhost:5000/waypoint-odr:dev"
+
+    use "exec" {
+      # executes a binary test with some arguments
+      command = "test"
+      args    = ["--full", "test-all"]
+    }
+  }
+}
+```
+
+### Embedded Pipelines in Steps
+
+You can optionally define a full pipeline within a Pipeline Step. This pattern
+is useful if you have a series of steps that make sense to be grouped together
+such as doing a deploy and release within a specific workspace.
+
+```hcl
+pipeline "simple-nested" {
+  step "build" {
+    use "build" {
+    }
+  }
+
+  step "deploy" {
+    pipeline "deploy" {
+      step "deploy" {
+        use "deploy" {
+        }
+      }
+      step "release" {
+        use "release" {
+          prune = true
+        }
+      }
+    }
+  }
+}
+```
+
+~> **NOTE**: Currently, Waypoint only allows for 1 level of nesting when defining an embedded pipeline. However, Pipelines can also be used as a reference within a step.
+
+### Reference Pipelines As Steps
+
+You can reference other Pipelines within a Project via a `"pipeline"` step:
+
+```hcl
+project = "my-pipeline-project"
+
+pipeline "simple-nested" {
+  step "build" {
+    use "build" {
+    }
+  }
+
+  step "my-deploy" {
+    #  "pipeline" is a special "built-in" step that references another pipeline
+    use "pipeline" {
+      project = "my-pipeline-project"
+      name    = "deploy"
+    }
+  }
+}
+
+pipeline "deploy" {
+  step "deploy" {
+    use "deploy" {
+    }
+  }
+  step "release" {
+    use "release" {
+      prune = true
+    }
+  }
+}
+```
+
+### Pipelines Across Workspaces
+
+TODO – Write about Waypoint Pipelines and how it works with workspaces
+
+Additionally, pipelines can be defined and scoped to a Workspace for deploying in various environments.
+
+```hcl
+
+```
+
+## Examples
+
+Below is an example that redefines the Up operation to be a pipeline with 3 steps for a build, deploy, and release. This example pipeline could be used as a starting point for more complicated pipelines.
+
+```hcl
+project = "my-project"
+
+pipeline "full-up" {
+  step "my-build" {
+    use "build" {
+    }
+  }
+
+  step "my-deploy" {
+    use "deploy" {
+    }
+  }
+
+  step "my-release" {
+    use "release" {
+    }
+  }
+}
+
+app "my-app" {
+  build { ... }
+  deploy { ... }
+  release { ... }
+}
+```
+
+For the “exec” steps, it accepts a command and arg parameter for running commands.
+
+```hcl
+pipeline "build-and-test" {
+  step "my-build" {
+    use "build" {
+    }
+  }
+  step "test" {
+    image_url = "https://localhost:5000/waypoint-odr:dev"
+    use "exec" {
+      # executes a binary test with some arguments
+      command = "test"
+      args    = ["--full", "test-all"]
+    }
+  }
+  step "my-deploy" {
+    use "deploy" {
+    }
+  }
+}
+```
+
+To run the "full-up" custom pipeline, it's as simple as `waypoint pipeline run full-up`:
```

Comment:
What does "full-up" mean?

Reply:
Oops, I think I had the pipeline name `full-up` in my notes but changed it later in the HCL example. Will fix!

Reply:
Oh wait, that is the name still, I can try to make this clearer.